timemachine/fe/system.py

```diff
@@ -157,15 +158,15 @@ def get_U_fns(self) -> List[BoundPotential[Potential]]:
 
 
 @dataclass
-class HostGuestSystem:
+class HostGuestSystem(Generic[_Nonbonded]):
```

Comment:
Is making the host guest system a nonbonded type required? This seems odd to make the host guest of the type nonbonded when it contains bonded potentials

Reply:
Note: this is not making the `HostGuestSystem` inherit from a nonbonded type. This is saying that `HostGuestSystem` is a generic type, parameterized by the type of the underlying nonbonded potential.

Reply:
I'm not sure what the :confused: emoji means.

Reply:
Just finding myself to dislike the python type hints more and more. But never invested any time to really understanding it. Just never know whether these additions are 'required', style or something else

Reply:
> Just never know whether these additions are 'required', style or something else

Type hints are never required; they are purely to enable more precise static checking of code. (In the case of dataclasses, a field types are syntactically required, but can be `Any`).

In this particular case making `HostGuestSystem` generic in `_Nonbonded` allows the type checker to keep track the type of the nonbonded potential that it contains. Weaker alternatives could include annotating `nonbonded_host_guest` as `Union[...]` or even `Any`.

Reply:
I get that they are optional, just when I see more exotic type hints like this one, it is difficult for me to discern the value.

For example, why is `Union[...]` weaker than setting up `HostGuestSystem` to be a generic on concrete types?

Reply:
@mcwitt clarified what is going offline, now makes sense!